### PR TITLE
Add basic ui tests for admin and user websites.

### PR DESCRIPTION
**Pull request by superhindupur**
## Purpose

* This PR adds Cypress UI tests - one for the user website and one for the admin website.
* These tests are not configured to run from Github actions yet - that will be covered in the next PR.
* Required by #419.

## Does this introduce a breaking change?

```
[ ] Yes
[x ] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[x] Other - Adds Cypress UI tests.
```

## How to Test

Open VSCode from within the dev container.
```
git clone https://github.com/Azure-Samples/chat-with-your-data-solution-accelerator.git
cd chat-with-your-data-solution-accelerator
git checkout bhindupur/ui-tests
npx cypress run --env BASE_URL_ADMIN_WEBSITE=https://example-admin.com,BASE_URL_USER_WEBSITE=https://example.com
```

## What to Check
Verify that the tests pass.

### CODE REVIEW

**Comment by ross-p-smith** [comment]
Hi @superhindupur - am I missing a step? I ran `npm install` beforehand and it asked me to install Cypress. 

```
❯ npx cypress run --env BASE_URL_ADMIN_WEBSITE=https://example-admin.com,BASE_URL_USER_WEBSITE=https://example.com
Need to install the following packages:
cypress@13.7.0
Ok to proceed? (y) y
It looks like this is your first time using Cypress: 13.7.0


Cypress failed to start.

This may be due to a missing library or dependency. https://on.cypress.io/required-dependencies

Please refer to the error below for more details.

----------

/home/vscode/.cache/Cypress/13.7.0/Cypress/Cypress: error while loading shared libraries: libatk-1.0.so.0: cannot open shared object file: No such file or directory

----------

Platform: linux-x64 (Debian - 11)
Cypress Version: 13.7.0
```

File: code/frontend/cypress.config.ts
Comment (ross-p-smith) [inline code/frontend/cypress.config.ts:1]: Wouldn't it be better to place these files in `/workspaces/chat-with-your-data-solution-accelerator/code/tests`

File: code/frontend/cypress/fixtures/ingest.txt
Comment (ross-p-smith) [inline code/frontend/cypress/fixtures/ingest.txt:1]: Appreciate that this is just a test file - but maybe place it in `/workspaces/chat-with-your-data-solution-accelerator/data`

File: code/frontend/cypress/support/e2e.ts
Comment (ross-p-smith) [inline code/frontend/cypress/support/e2e.ts:1]: This looks like it just imports a commented out file - do we need either of these?

File: code/frontend/cypress.config.ts
Comment (adamdougal) [reply code/frontend/cypress.config.ts:1]: Maybe `code/tests/e2e` to follow the same pattern that we have with the functional tests.

File: code/frontend/cypress/e2e/admin.spec.cy.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+describe('the cwyd admin website', () => {
+  before(() => {
+    const myVariable = Cypress.env('BASE_URL_ADMIN_WEBSITE');
+    cy.log('My variable:', myVariable);
+    cy.visit('/Ingest_Data');
+  });
+
+  it('allows file upload', () => {
+    cy.get('input[type=file]', { timeout: 30000 }).selectFile('cypress/fixtures/ingest.txt', { force: true });
```
Comment (adamdougal) [inline code/frontend/cypress/e2e/admin.spec.cy.ts:11]: Should we add a verification that the processing/upload successful message is then displayed?

File: code/tests/integration/ui/cypress/README.md [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+Run the command for cypress tests (to run in headless mode)
+
+```
+npx cypress run --env BASE_URL_ADMIN_WEBSITE=https://example-admin.com,BASE_URL_USER_WEBSITE=https://example.com
+```
+
+If you want to run the tests in interactive mode (in a browser)
+
+```
+npx cypress --env BASE_URL_ADMIN_WEBSITE=https://example-admin.com,BASE_URL_USER_WEBSITE=https://example.com
```
Comment (cecheta) [inline code/tests/integration/ui/cypress/README.md:12]:
```suggestion
npx cypress open --env BASE_URL_ADMIN_WEBSITE=https://example-admin.com,BASE_URL_USER_WEBSITE=https://example.com
```

I think this is missing?

File: package-lock.json
Comment (cecheta) [inline package-lock.json:1]: Which `package.json` is this file related to?

File: code/tests/integration/ui/package.json
Comment (cecheta) [inline code/tests/integration/ui/package.json:1]: We should add this file to the dependabot config

File: code/tests/integration/ui/cypress/support/e2e.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
+// ***********************************************************
```
Comment (cecheta) [inline code/tests/integration/ui/cypress/support/e2e.ts:1]: I see that others have already mentioned these commented-out files, but seems like we can disable from the config the importing of these files if we're not using them

File: code/frontend/cypress.config.ts
Comment (superhindupur) [reply code/frontend/cypress.config.ts:1]: Put the tests in code/tests/integration/ui based on our chat later (where we discussed we will have UI and API based e2e tests, with this PR introducing UI tests)

File: code/frontend/cypress/fixtures/ingest.txt
Comment (superhindupur) [reply code/frontend/cypress/fixtures/ingest.txt:1]: Done, I just reused a file already present in /data

File: code/frontend/cypress/support/e2e.ts
Comment (superhindupur) [reply code/frontend/cypress/support/e2e.ts:1]: Removed

File: code/frontend/cypress/e2e/admin.spec.cy.ts (hunk 1 above)
Comment (superhindupur) [reply code/frontend/cypress/e2e/admin.spec.cy.ts]: Done

File: code/frontend/cypress/fixtures/ingest.txt [hunk 4]
```diff
@@ -0,0 +1 @@
+Microsoft Corporation is an American multinational corporation and technology company headquartered in Redmond, Washington.[2] Microsoft's best-known software products are the Windows line of operating systems, the Microsoft 365 suite of productivity applications, and the Edge web browser. Its flagship hardware products are the Xbox video game consoles and the Microsoft Surface lineup of touchscreen personal computers. Microsoft ranked No. 14 in the 2022 Fortune 500 rankings of the largest United States corporations by total revenue;[3] and it was the world's largest software maker by revenue in 2022 according to Forbes Global 2000. It is considered one of the Big Five American information technology companies, alongside Alphabet (parent company of Google), Amazon, Apple, and Meta (parent company of Facebook).
```
[the remark this answers is not in the document]
Comment (superhindupur) [reply code/frontend/cypress/fixtures/ingest.txt]: N/A since I deleted this file. I am now using an existing file from/data

File: code/frontend/cypress/support/commands.ts [hunk 5]
```diff
@@ -0,0 +1,37 @@
+/// <reference types="cypress" />
```
[the remark this answers is not in the document]
Comment (superhindupur) [reply code/frontend/cypress/support/commands.ts]: Removed it now

File: code/tests/integration/ui/package.json
Comment (superhindupur) [reply code/tests/integration/ui/package.json:1]: Done

File: package-lock.json
Comment (superhindupur) [reply package-lock.json:1]: This is now removed, it was added in error. Thanks!

File: .devcontainer/postCreate.sh [hunk 6]
```diff
@@ -6,4 +6,8 @@ pip install poetry
 
 poetry install --with dev
 
-poetry run pre-commit install
\ No newline at end of file
+poetry run pre-commit install
+
+sudo apt-get update
+
+sudo apt-get install libgtk2.0-0 libgtk-3-0 libgbm-dev libnotify-dev libnss3 libxss1 libasound2 libxtst6 xauth xvfb
```
Comment (cecheta) [inline .devcontainer/postCreate.sh:13]: Could this be in the Dockerfile rather than here?

Comment (ross-p-smith) [inline .devcontainer/postCreate.sh:13]: This needs to go in the docker file and not here (also you shouldn't need sudo at that point)

File: azure.yaml
Comment (cecheta) [inline azure.yaml:1]: I don't think we can make these changes as we are still deploying from `./code`

File: Makefile [hunk 7]
```diff
@@ -13,6 +13,10 @@ ifeq ($(filter $(MAKECMDGOALS),config clean),)
 	endif
 endif
 
+AZURE_ENV_FILE := $(shell azd env list --output json | jq -r '.[] | select(.IsDefault == true) | .DotEnvPath')
```
Comment (ross-p-smith) [inline Makefile:16]: Maybe this should be above the `ENV_FILE := .env` line and if it errors, it defaults to `.env` so if someone hasn't configured azd yet that everything fails gracefully. You could then feed the filename into the existing logic that includes and exports - just a thought

**Comment by ross-p-smith** [comment]
I think you need to rebase ontop of Ren's changes so you can get the docker file you need.

**Comment by ross-p-smith** [comment]
I can get this to run - I think the final must have is moving the postCreate into the Dockerfile
image

File: Makefile [hunk 8]
```diff
@@ -30,6 +34,10 @@ functionaltest: ## 🧪 Run the functional tests
 	@echo -e "\e[34m$@\e[0m" || true
 	@ poetry run pytest -m "functional"
 
+uitest: ## 🧪 Run the ui tests in headless mode
+	@echo -e "\e[34m$@\e[0m" || true
+	@cd code/tests/integration/ui && npm install && npx cypress run --env ADMIN_WEBSITE_NAME=$(ADMIN_WEBSITE_NAME),FRONTEND_WEBSITE_NAME=$(FRONTEND_WEBSITE_NAME)
```
[the remark this answers is not in the document]
Comment (superhindupur) [reply Makefile:39]: Cypress only reads OS environment variables directly if they are prefixed with CYPRESS_, which I don't think we want to do with our outputted env vars since it will tie them to Cypress via their names.

The child shell is able to read the variables, which is why  $(ADMIN_WEBSITE_NAME) is working - but we just need to explicitly pass them on to cypress via the CLI because they aren't prefixed with CYPRESS_.

Reference: https://docs.cypress.io/guides/guides/environment-variables#Setting

File: Makefile (hunk 7 above)
Comment (superhindupur) [reply Makefile:16]: The ui tests will not run successfully unless the azd env is configured anyway. The failure is graceful already - it just ignores the including of the azure env file and goes ahead to run the tests. If the env variables are available from any source, the cypress command will pick them up.

File: .devcontainer/postCreate.sh [hunk 9]
```diff
@@ -6,4 +6,4 @@ pip install poetry
 
 poetry install --with dev
 
-poetry run pre-commit install
```
Comment (ross-p-smith) [inline .devcontainer/postCreate.sh:9]: Sounds petty - but can we add the new line back in. Newlines are useful at the bottom of files

**Comment by superhindupur** [comment]
Since we eventually want to move all tests outside the code/ folder, I took the first step by placing the UI tests at tests/integration/ui. This means that the azd build config in azure.yaml can stay as is.

**Review by ross-p-smith — APPROVED** [review]
All works. #541 will fix the folder structure going forward

**Review by cecheta — APPROVED** [review]
Great stuff :)
